### PR TITLE
Update BattleArena Action

### PR DESCRIPTION
**Pull request by hskim881028**
- Target version : `v100240`
- Add ticket purchase function
- Fixed to only allow a single battle during the season
- Related link:
   -  https://www.notion.so/planetarium/4-d41f04d0afdf44f8b51e9250c5fa4528
   -  https://planetariumhq.slack.com/archives/C03BU3JTKAN/p1656466364090499
   - image

### CODE REVIEW

File: Lib9c/Arena/ArenaHelper.cs [hunk 1]
```diff
@@ -136,5 +137,18 @@ public static int GetRewardCount(int score)
...
+        public static FungibleAssetValue GetTicketPrice(
+            ArenaSheet.RoundData roundData,
+            ArenaInformation arenaInformation,
+            Currency currency)
+        {
+            var price = roundData.TicketPrice * 0.01m
+                        + (roundData.AdditionalTicketPrice * 0.01m * arenaInformation.PurchasedTicketCount);
+            var major = (long)price;
+            var miner = (int)((price - major) * 100m);
+            var ticketPrice = new FungibleAssetValue(currency, major, miner);
```
Comment (ipdae) [inline Lib9c/Arena/ArenaHelper.cs:150]: we can use `FungibleAssetValue.DivRem` like this. https://github.com/planetarium/lib9c/blob/65f8acdf46a54196f83816d42933804483a3d4ab/Lib9c/Helper/CrystalCalculator.cs#L102

File: Lib9c/Action/BattleArena.cs [hunk 2]
```diff
@@ -223,7 +224,27 @@ public override IAccountStateDelta Execute(IActionContext context)
...
+            {
+                if (ticket > 1)
+                {
+                    throw new ExceedTicketPurchaseCountException($"[{nameof(BattleArena)}] ticket : {ticket}");
+                }
+
+                var ticketBalance = ArenaHelper.GetTicketPrice(roundData, arenaInformation, states.GetGoldCurrency());
+                var arenaAdr = ArenaHelper.DeriveArenaAddress(roundData.ChampionshipId, roundData.Round);
+                states = states.TransferAsset(context.Signer, arenaAdr, ticketBalance);
+                arenaInformation.BuyTicket();
```
Comment (ipdae) [inline Lib9c/Action/BattleArena.cs:246]: is this correct? i can't found use ticket in this case.

Comment (hskim881028) [reply Lib9c/Action/BattleArena.cs]: We can only buy one ticket. So I didn't create any parameters in Buy Ticket()

File: Lib9c/Arena/ArenaHelper.cs [hunk 3]
```diff
@@ -136,5 +137,23 @@ public static int GetRewardCount(int score)
...
             return 1;
         }
+
+        public static FungibleAssetValue GetTicketPrice(
+            ArenaSheet.RoundData roundData,
+            ArenaInformation arenaInformation,
+            Currency currency)
+        {
+            var ticketPrice = new FungibleAssetValue(currency, roundData.TicketPrice, 0);
+            var addTicketPrice = new FungibleAssetValue(currency, roundData.AdditionalTicketPrice, 0);
```
Comment (ipdae) [inline Lib9c/Arena/ArenaHelper.cs:147]:
```suggestion
            var ticketPrice = currency * roundData.TicketPrice;
            var addTicketPrice = currency * roundData.AdditionalTicketPrice;
```

File: .Lib9c.Tests/Action/ArenahelperTest.cs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    using Serilog;
+    using Xunit;
+    using Xunit.Abstractions;
+    using static Lib9c.SerializeKeys;
+
+    public class ArenaHelperTest
+    {
+        private IAccountStateDelta _state;
+        private Currency _crystal;
+        // private readonly Currency _ncg;
```
Comment (ipdae) [inline .Lib9c.Tests/Action/ArenahelperTest.cs:27]:
```suggestion
```

File: Lib9c/Action/BattleArena.cs [hunk 5]
```diff
@@ -223,7 +224,26 @@ public override IAccountStateDelta Execute(IActionContext context)
...
+            if (arenaInformation.Ticket > 0)
+            {
+                arenaInformation.UseTicket(ticket);
+            }
+            else
+            {
+                for (var i = 0; i < ticket; i++)
+                {
+                    var ticketBalance = ArenaHelper.GetTicketPrice(roundData, arenaInformation, states.GetGoldCurrency());
+                    var arenaAdr = ArenaHelper.DeriveArenaAddress(roundData.ChampionshipId, roundData.Round);
```
Comment (ipdae) [inline Lib9c/Action/BattleArena.cs:242]: i think it can be called once. because address can't change in this loop. right?

Comment (hskim881028) [reply Lib9c/Action/BattleArena.cs]: 👍🏻 I fixed it 🙏🏻 @ipdae

**Review by boscohyun — APPROVED** [review]

**Pull request merged**